subsys/net/lib/app/client.c

```diff
@@ -0,0 +1,603 @@
+		/* If we have TLS connection, the connect cb is called
+		 * after TLS handshakes are done.
+		 */
+		NET_DBG("Postponing TLS connection cb for ctx %p", ctx);
+	} else
+#endif
+	{
+		if (ctx->cb.connect) {
+			ctx->cb.connect(ctx, status, ctx->user_data);
+		}
+	}
+}
+
+int net_app_connect(struct net_app_ctx *ctx, s32_t timeout)
+{
+	struct net_context *net_ctx;
+	bool started = false;
+	int ret;
+
+	if (!ctx) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -ENOENT;
+	}
+
+	if (ctx->app_type != NET_APP_CLIENT) {
+		return -EINVAL;
+	}
+
+	net_ctx = _net_app_select_net_ctx(ctx);
+	if (!net_ctx) {
+		return -EAFNOSUPPORT;
+	}
+
+#if defined(CONFIG_NET_APP_TLS)
+	if (ctx->is_tls && ctx->proto == IPPROTO_TCP && !ctx->tls.tid) {
+		/* TLS thread is not yet running, start it now */
+		ret = start_tls_client(ctx);
+		if (ret < 0) {
+			NET_DBG("TLS thread cannot be started (%d)", ret);
+			return ret;
+		}
+
+		started = true;
+
+		/* Let the TLS thread run first */
+		k_yield();
+	}
+#else
+	ARG_UNUSED(started);
+#endif /* CONFIG_NET_APP_TLS */
+
+	ret = net_context_connect(net_ctx,
+				  &ctx->remote,
+				  sizeof(ctx->remote),
+				  _app_connected,
+				  timeout,
+				  ctx);
+	if (ret < 0) {
+		NET_DBG("Cannot connect to peer (%d)", ret);
+
+#if defined(CONFIG_NET_APP_TLS)
+		if (started) {
+			_net_app_tls_handler_stop(ctx);
+		}
+#endif
+	}
+
+	return ret;
+}
+
+#if defined(CONFIG_NET_APP_TLS)
+static void tls_client_handler(struct net_app_ctx *ctx,
+			       struct k_sem *startup_sync)
+{
+	int ret;
+
+	NET_DBG("Starting TLS client thread for %p", ctx);
+
+	ret = _net_app_tls_init(ctx, MBEDTLS_SSL_IS_CLIENT);
+
+	k_sem_give(startup_sync);
+
+	if (ret < 0) {
+		NET_DBG("TLS client init failed");
+		return;
+	}
+
+	/* We wait until TLS connection is established */
+	k_sem_take(&ctx->client.connect_wait, K_FOREVER);
+
+	_net_app_ssl_mainloop(ctx);
+
+	mbedtls_ssl_close_notify(&ctx->tls.mbedtls.ssl);
+
+	/* If there is any pending data that have not been processed
+	 * yet, we need to free it here.
+	 */
+	if (ctx->tls.mbedtls.ssl_ctx.rx_pkt) {
+		net_pkt_unref(ctx->tls.mbedtls.ssl_ctx.rx_pkt);
+		ctx->tls.mbedtls.ssl_ctx.rx_pkt = NULL;
+		ctx->tls.mbedtls.ssl_ctx.frag = NULL;
+	}
+
+	if (ctx->cb.close) {
+		ctx->cb.close(ctx, -ESHUTDOWN, ctx->user_data);
+	}
+
+	_net_app_tls_handler_stop(ctx);
+}
+
+static int start_tls_client(struct net_app_ctx *ctx)
+{
+	struct k_sem startup_sync;
+
+	/* Start the thread that handles TLS traffic. */
+	if (ctx->tls.tid) {
+		return -EALREADY;
+	}
+
+	k_sem_init(&startup_sync, 0, 1);
+
+	ctx->tls.tid = k_thread_create(&ctx->tls.thread,
+				       ctx->tls.stack,
+				       ctx->tls.stack_size,
+				       (k_thread_entry_t)tls_client_handler,
+				       ctx, &startup_sync, 0,
+				       K_PRIO_COOP(7), 0, 0);
+
+	/* Wait until we know that the TLS thread startup was ok */
+	if (k_sem_take(&startup_sync, TLS_STARTUP_TIMEOUT) < 0) {
+		_net_app_tls_handler_stop(ctx);
```

Comment:
If mbedtls stuff will be released in _net_app_tls_init() due to some error, we should avoid to call_net_app_tls_handler_stop() here.